src/klein/_app.py

```diff
@@ -53,11 +80,28 @@ def _call(__klein_instance__, __klein_f__, *args, **kwargs):
 @implementer(IKleinRequest)
 class KleinRequest(object):
     def __init__(self, request):
+        # type: (Request) -> None
         self.branch_segments = [""]
-        self.mapper = None
 
-    def url_for(self, *args, **kwargs):
-        return self.mapper.build(*args, **kwargs)
+        # Don't annotate as optional, since you should never set this to None
```

Comment:
I wanted to try to express this as you've stated it with a property, but there appears to be a bug in mypy:

https://github.com/python/mypy/issues/9180

The real problem here though appears to be that we are constructing a KleinRequest via adaptation - apparently for no good reason.  Rather than lie about the type of `mapper`, I think it would make more sense to replace `KleinResource.render`'s invocation with a simple `KleinRequest(request, mapper)`.

Technically what I'm proposing is an incompatible change, but given that `KleinRequest(request)` gives you a nonsense non-functional object without an attached mapper, I'm OK with a short deprecation cycle here.

So I think adding types here has revealed a very real bug :)

Reply:
https://github.com/twisted/klein/issues/378